text/0000-deprecation.md

```diff
@@ -0,0 +1,111 @@
+- Feature Name: Public Stability
+- Start Date: 2015-09-03
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+
+This RFC proposes to allow library authors to use a `#[deprecated]` attribute,
+with optional `since = "`*version*`"`, `reason = "`*free text*`"` and 
+`use = "`*substitute declaration*`"` fields. The compiler can then
+warn on deprecated items, while `rustdoc` can document their deprecation
+accordingly. 
+
+# Motivation
+
+Library authors want a way to evolve their APIs; which also involves 
+deprecating items. To do this cleanly, they need to document their intentions 
+and give their users enough time to react.
+
+Currently there is no support from the language for this oft-wanted feature
+(despite a similar feature existing for the sole purpose of evolving the Rust
+standard library). This RFC aims to rectify that, while giving a pleasant
+interface to use while maximizing usefulness of the metadata introduced.
+
+# Detailed design
+
+Public API items (both plain `fn`s, methods, trait- and inherent 
+`impl`ementations as well as `const` definitions, type definitions, struct
+fields and enum variants) can be given a `#[deprecated]` attribute. All
+possible fields are optional:
+
+* `since` is defined to contain the exact version of the crate that
+deprecated the item, as defined by Cargo.toml (thus following the semver
+scheme). It makes no sense to put a version number higher than the current
+newest version here, and this is not checked (but could be by external
+lints, e.g. [rust-clippy](https://github.com/Manishearth/rust-clippy).
+To maximize usefulness, the version should be fully specified (e.g. no
+wildcards or ranges).
+* `reason` should contain a human-readable string outlining the reason for
+deprecating the item. While this field is not required, library authors are
+strongly advised to make use of it to convey the reason to users of their
+library. The string is required to be plain unformatted text (for now) so that
+rustdoc can include it in the item's documentation without messing up the 
+formatting.
+* `use` should be the full path to an API item that will replace the 
+functionality of the deprecated item, optionally (if the replacement is in a 
+different crate) followed by `@` and either a crate name (so that 
+`https://crates.io/crates/` followed by the name is a live link) or the URL to 
+a repository or other location where a surrogate can be obtained. Links must be 
+plain FTP, FTPS, HTTP or HTTPS links. The intention is to allow rustdoc (and
+possibly other tools in the future, e.g. IDEs) to act on the included 
+information. The `use` field can have multiple values.
```

Comment:
You're right about that. I'm changing this to just use paths (like import paths) – let's just restrict the search to crates on the path. This means we can easily check if the item exists and is public.